Use a "type structure" of each impl to choose the best match

danakj opened the pull request:
The type structure is built for the impl lookup query for the combination of the self type and the interface being queried. Then it is built for each `impl` definition that is a potential candidate.

The type structures are compared to ensure they have a compatible structure, and the `impl` declaration is not considered if they do not.

Finally, the type structures are used as a sorting key for the candidate `impl` declarations, with the most-specified type structures (the ones with the furthest distance to the first symbolic value) coming first in the ordering.

See https://docs.carbon-lang.dev/docs/design/generics/overview.html#parameterized-impl-declarations for the design of the type structure and related ordering.

Most of the commits in this PR landed in #5158 (11ae0e27ab5c121) by mistake, but this includes the final changes since that PR was written.

danakj committed:
Use a "type structure" of each impl to choose the best match

The type structure is not used for throwing out impls yet, but it is
used to order them from best to worst match, and the best match is
chosen. This allows a specialized impl to be used over a more general
one.

danakj commented:
If preferred I can work on matching impls against type structures before landing this. This PR doesn't do anything like that and the type structure has none of the state that would be needed for it yet.

danakj commented:
Rebased onto https://github.com/carbon-language/carbon-lang/pull/5126

edit: rebased back to trunk

danakj committed:
Build the full type structure for an impl

Keep traversing type structure after finding first symbolic, and add
concrete/symbolic items to a structural representation in a vector. When
we find a type with testing, we mark the opening/closing of that type so
that a symbolic will know which things it matches against. If it's
positioned at the same place as an concrete opening, then it matches
everything up to and including the paired concrete closing.

Matching with this structure is not implemented yet. Everything just
returns true.

danakj committed:
add import test with GenericClass as parameter

danakj committed:
matching

danakj commented:
Since this isn't reviewed yet, I just pushed all the specialization work into the same PR.

danakj committed:
dump

danakj committed:
autoupdate

zygoloid commented on toolchain/check/testdata/function/generic/deduce.carbon:
```diff
@@ -191,6 +191,38 @@ fn CallImplicitNotDeducible() {
+impl forall [E:! type] CC(DD(E)) as Z {}
+
+fn F() {
+  // TODO: Deduction of `E` in the impl of `CC` as `Z` fails with:
```
I guess the problem is that `FacetValue` doesn't set `.deduce_through = true`, so we can't deduce the `FacetValue(DD(E), <witness>)` against `FacetValue(DD(EE), <witness>)`?

danakj replied:
Woah, yes that is correct.

I have a follow-up PR that will fix this then.

zygoloid commented on toolchain/check/testdata/impl/lookup/min_prelude/impl_cycle.carbon:
```diff
@@ -19,36 +19,44 @@ interface Z {}
+  // CHECK:STDERR: fail_impl_simple_cycle.carbon:[[@LINE+4]]:3: note: type `type` does not implement interface `Core.As(Z)` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   Point as Z;
+  // CHECK:STDERR:   ^~~~~~~~~~
+  // CHECK:STDERR:
```
Would it be worth adding a TODO here to somehow suppress the second diagnostic?

danakj replied:
Yeah. I think this will fall out as part of the change to making impl lookup errors into notes for type-value/facet conversions. Cuz then the first diagnostic will actually become a note of the second, I guess. But maybe for a cycle we make a different choice, either way, done.

zygoloid commented on toolchain/check/testdata/impl/lookup/min_prelude/impl_cycle.carbon:
```diff
@@ -60,20 +68,25 @@ interface Y {}
+class Point {}
 
 fn F() {
   // TODO: A cycle should be found in the `impl forall T as Z` above.
```
Looks like this TODO is done already.

danakj replied:
Yes!

zygoloid commented on toolchain/check/testdata/impl/lookup/min_prelude/specialization.carbon:
```diff
@@ -0,0 +1,274 @@
+fn F() {
+  // The specialization of `Z(C)` should match in preference to the blanket impl
+  // of `Z(T)`. If the blanket impl is chosen, then `a` will have type `()`
+  // which will fail to typecheck here.
```
```suggestion
  // The specialization of `C(())` should match in preference to the blanket impl
  // of `C(T)`. If the blanket impl is chosen, then `a` will have type `()`
  // which will fail to typecheck here.
```

zygoloid commented on toolchain/check/testdata/impl/lookup/min_prelude/specialization.carbon:
```diff
@@ -0,0 +1,274 @@
+  // of `Z(T)`. If the blanket impl is chosen, then `a` will have type `()`
+  // which will fail to typecheck here.
+  //
+  // TODO: C(T) is C(()) so this should work?
```
You could work around this bug by using a non-symbolic value for the associated type in the expected impl, eg `C(())` instead of `C(T)`.

danakj replied:
Oh yes good idea, thanks.

zygoloid commented on toolchain/check/type_structure.h:
```diff
@@ -0,0 +1,105 @@
+               rhs.distance_to_first_symbolic_type_;
+      }
+    }
+    return std::weak_ordering::equivalent;
```
Maybe include a TODO here to consider the positions of remaining symbolic types?

danakj replied:
Oh, right. Perhaps just tracking the first one isn't so helpful then - though it's certainly quicker for comparison. Left the TODO.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+      return false;
+    }
+    // Same structural element on both sides, they match and both are consumed.
+    if (*lhs_cursor == *rhs_cursor) {
```
Maybe include a TODO to track the concrete value for `Structural::Concrete` so we can compare it here?

danakj replied:
Nice idea, left the TODO.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+    // and make sure the cursors are swapped back at the bottom of the loop
+    // iteration, so that they compare with the correct `end()`s.
+    auto unswap =
+        llvm::make_scope_exit([&] { std::swap(lhs_cursor, rhs_cursor); });
```
I wonder if this would be simpler if you factor the rest of this out to a function / lambda and pass in either `lhs_cursor, rhs_cursor` or `rhs_cursor, lhs_cursor` by reference.

danakj replied:
In one way, yes. But the lambda will need a multi-state return that indicates if we should `continue` or `return false` after its own return, and that works with references to pointers (iterators). I've made it use a lambda - do you like that more than the scoped swap? I weakly prefer the swap, I think.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+    if (*lhs_cursor == Structural::Concrete) {
+      ++lhs_cursor;
+      continue;
+    }
```
This check looks redundant to me (the loop below does the same thing). I'm fine with keeping this if you think it makes the flow here clearer, but I find it clear enough that the loop below will consume one non-bracket token or one balanced bracketed group.

danakj replied:
Ah.. yeah the loop used to not be redundant, when it was written slightly different, but I guess it is now. :) I will drop the if here.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+}
+
+// A class that builds a `TypeStructure` for an `Impl` that represents its self
+// type.
```
```suggestion
// type and interface.
```

danakj replied:
Done

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+      work_list_.pop_back();
+
+      if (std::holds_alternative<CloseType>(next)) {
+        auto close_type = std::get<CloseType>(next);
```
This can be done more compactly:
```suggestion
      if (auto* close_type = std::get_if<CloseType>(&next)) {
```

danakj replied:
Done in a couple places.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
```
Consider merging these identical cases:
```suggestion
        case SemIR::AssociatedEntityType::Kind:
        case SemIR::BoolType::Kind:
        case SemIR::FloatType::Kind:
        case SemIR::GenericClassType::Kind:
        case SemIR::GenericInterfaceType::Kind:
        case SemIR::ImplWitnessAccess::Kind: {
          ++position_;
          structure_.push_back(TypeStructure::Structural::Concrete);
          break;
        }
```
and similar.

danakj replied:
Oh, I didn't realize I can do that in a CARBON_KIND_SWITCH, thanks!

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+
+  Context& context_;
+  llvm::SmallVector<WorkItem> work_list_;
+  int position_;
```
I've not checked carefully, but it looks like `position_` might always be equal to `structure_.size()`.

danakj replied:
You are correct, yeah. To make the code clear and avoid mistakes, I have added a helper function to append elements to `structure_`, and it also records the first symbolic element's position. I called this operation "append" to differentiate it from "push" which is used to add work to the stack.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+          structure_.push_back(TypeStructure::Structural::ConcreteOpenParen);
+          Push(CloseType{});
+          Push(array_type.element_type_id);
+          break;
```
Do we also need to push the width? That might contain a `?`.

danakj replied:
Oh right, yes

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+          Push(CloseType{.closing = facet_type_info.other_requirements
+                                        ? CloseWithSymbolic
+                                        : CloseWithConcrete});
+          for (const auto& i : facet_type_info.impls_constraints) {
```
Hm, `impls_constraints` are sorted in a sort of arbitrary order here. That seems problematic! 

Actually -- I think perhaps we shouldn't be recursing into the constraints of a facet type at all. We don't deduce through them, and I think we shouldn't have `?`s in the type structure in non-deduced positions.

danakj replied:
Right, I have also convinced myself that it is not necessary or correct to look at the interfaces in a FacetType, and I have written a comment explaining it.

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+      inst_id = facet_value->type_inst_id;
+    }
+
+    auto type_id_of_inst_id = context_.insts().Get(inst_id).type_id();
+    if (context_.types().Is<SemIR::FacetType>(type_id_of_inst_id)) {
+      return SymbolicType();
+    }
+    if (type_id_of_inst_id != SemIR::TypeType::SingletonTypeId) {
+      return SemIR::TypeId::None;
+    }
```
A comment explaining why it's correct to treat non-type facets as always being symbolic would be useful. (I think this might be right, but I'm not certain.)

danakj replied:
Done

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -0,0 +1,455 @@
+    // the queue works from the back.
+    Push(interface_constraint);
+
+    auto maybe_self_type_id = TryGetInstIdAsTypeId(self_inst_id);
+    if (std::holds_alternative<SemIR::TypeId>(maybe_self_type_id)) {
+      auto self_type_id = std::get<SemIR::TypeId>(maybe_self_type_id);
+      CARBON_CHECK(self_type_id != SemIR::TypeId::None);
+      Push(self_type_id);
+    } else {
+      CARBON_CHECK(std::holds_alternative<SymbolicType>(maybe_self_type_id));
+      Push(SymbolicType());
+    }
```
It looks like this duplicates some of the work of `PushArgs`; would it make sense to factor out a `PushArg`?

danakj replied:
Ah, yes, and I need the same for pushing the array bound, which is an instruction id not a type id.

danakj committed:
todo-diagnostic

danakj committed:
better-comment

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

danakj committed:
push array bound in type structure

danakj committed:
Remove position_ in type structure builder

The position is the size of the structure, now that we are building
the full structure. To help make this all more clear we move the
pushing into the type structure vector into a helper method and
have it record the position of the first symbolic appended to the
type structure.

We use 'append' for the operation of adding to the type structure
to try differentiate it from 'push' which is used to add work to
the work stack.

danakj committed:
Add deduce test case with a deduced FacetValue in a class parameter

The deduced FacetValue restricts what the impl matches even though the
caller has specified a type that should match the impl's generic
constraints because the class's FacetValue constrains it further.

danakj committed:
Comment to explain why all FacetType-typed insts are symbolic

This would not be true if we included FacetValues which may or may not
be symbolic, but they are handled earlier and converted into an inst
of type TypeType, not FacetType.

danakj committed:
FacetType is concrete

danakj committed:
Add test with symbolic facet value

danakj committed:
Work around associated constraints issue, use a concrete type in rewrite

danakj committed:
autoupdate

danakj committed:
Merge remote-tracking branch 'origin/trunk' into specialize

danakj committed:
autoupdate

danakj committed:
Format

danakj reviewed (commented):
Thanks! PTAL

danakj commented:
Oh github had hidden some comments from me in convo view, let me address those first.

danakj committed:
lambda-instead-of-swap

danakj committed:
todos

danakj committed:
drop-if-concrete

danakj committed:
comment

danakj committed:
get_if

danakj committed:
Remove CloseType field. Add comments.

danakj committed:
switch-kind-compact

danakj committed:
pointers

danakj committed:
use ()

danakj commented:
Ok this is ready for review again.

Reading the design I saw it mention `Ptr(T)` for pointers in the type structure and realized I was not chasing through pointers. I added a test that failed and fixed it by making pointers into an aggregate type, over the pointee type, in the structure.

zygoloid commented on toolchain/check/testdata/function/generic/deduce.carbon:
```diff
@@ -223,6 +223,50 @@ fn F() {
 }
 
+// --- fail_deduce_nested_facet_value.carbon
+// CHECK:STDERR: fail_deduce_nested_facet_value.carbon: error: `Main//default` previously provided by `fail_todo_deduce_nested_generic_class.carbon` [DuplicateMainApi]
```
Missing `library` declaration here (and in the other split).

danakj replied:
Done

zygoloid commented on toolchain/check/testdata/function/generic/deduce.carbon:
```diff
@@ -223,6 +223,50 @@ fn F() {
+// constructed by deducing CC(D) has a specific argument D that only
+// satisfies Y. So CC(DD), perhaps surprisingly from the call site, does not
+// match this impl declaration. The diagnostic indicates that the type being
+// converted is `CC(DD as Y)` rather than `CC(DD)`.
```
From open discussion today, I think the outcome we want here is to allow this. Maybe add a TODO?

danakj replied:
Done

danakj committed:
Make deduce type from nested FacetValue test into a fail_todo_ and leave a TODO

zygoloid commented on toolchain/check/testdata/impl/lookup/min_prelude/specialization.carbon:
```diff
@@ -251,6 +217,69 @@ fn F() {
   let a: C(D(E), D(E)).(Z.X) = {} as C(E, E);
 }
 
+// --- specialed_class_with_symbolic_facet_value_param.carbon
```
```suggestion
// --- specialized_class_with_symbolic_facet_value_param.carbon
```

zygoloid commented on toolchain/check/testdata/impl/lookup/min_prelude/specialization.carbon:
```diff
@@ -251,6 +217,69 @@ fn F() {
+  // The FacetValue deduced for the param of `C` will be a symbolic FacetValue
+  // because we are in a generic where `D` is an unknown type, which will cause
+  // the query and impl self type to be C(FacetValue) for a symbolic FacetValue.
+  let a: C(D).(Z.X) = ();
```
Because the `impl forall [D:! Y] ...` is not final, I think `C(D).(Z.X)` here should be symbolic and should not resolve to `()`. But we have no support for `final impl`s yet, so maybe add a TODO?

I think the case where we pick the impl eagerly despite it not being final is when the query is concrete. In that case we can't delay the query until the enclosing generic is instantiated (because there isn't one), so instead we'd commit early to using the parameterized impl and we'd need to reject if a more specific impl is given later.

danakj replied:
Done

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -48,64 +51,67 @@ auto TypeStructure::IsCompatibleWith(const TypeStructure& other) const -> bool {
     // From here we know one side is a Symbolic and the other is not. We can
     // match the Symbolic against either a single Concrete or a larger bracketed
     // set of Concrete structural elements.
```
Optional: instead of checking for one of the two operands being symbolic here, maybe integrate this into the `if`s on line 99-107.

danakj replied:
I think I prefer to keep it above the lambda, marking the section of code below as being about this, even if the next lines are not really what runs next.

danakj committed:
Update toolchain/check/testdata/impl/lookup/min_prelude/specialization.carbon

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

danakj committed:
Merge branch 'trunk' into specialize

danakj committed:
Add todo to add the type for non-type values

danakj committed:
Add todo about symbolic witness results

danakj committed:
autoupdate

danakj committed:
handle TypeOfInst

danakj commented:
There's a new `TypeOfInst` instruction that comes up in the type structure now.

For now, I am just pushing the `.inst_id` from the instruction to look through it. The thing inside it in this test is a `AccessMemberAction` which is not a type value, so it results in a `Concrete` in the structure as a `NonTypeValue`.

I am not super convinced if this is correct.. but if we are dependent on a template type I guess we won't be using that type to do impl lookup as a symbolic right? We'd be waiting until there's a specific, so maybe TypeOfInst should just be Concrete without worrying about what is inside?

danakj commented:
> There's a new `TypeOfInst` instruction that comes up in the type structure now.
> 
> For now, I am just pushing the `.inst_id` from the instruction to look through it. The thing inside it in this test is a `AccessMemberAction` which is not a type value, so it results in a `Concrete` in the structure as a `NonTypeValue`.
> 
> I am not super convinced if this is correct.. but if we are dependent on a template type I guess we won't be using that type to do impl lookup as a symbolic right? We'd be waiting until there's a specific, so maybe TypeOfInst should just be Concrete without worrying about what is inside?

Hm, the [doc](https://docs.google.com/document/d/1eWW8MTko3PIqxZ32-GhsdaRSYqoDicxMB1VeessMTOg/edit?pli=1&tab=t.0) says of `Action`s: "Whether a constant or type is template-dependent is tracked on its constant value." So I guess TypeOfInst can have an action that is not a template.

Perhaps I just need to look at its value and see if it's template/symbolic or not.

danakj commented:
> > There's a new `TypeOfInst` instruction that comes up in the type structure now.
> > For now, I am just pushing the `.inst_id` from the instruction to look through it. The thing inside it in this test is a `AccessMemberAction` which is not a type value, so it results in a `Concrete` in the structure as a `NonTypeValue`.
> > I am not super convinced if this is correct.. but if we are dependent on a template type I guess we won't be using that type to do impl lookup as a symbolic right? We'd be waiting until there's a specific, so maybe TypeOfInst should just be Concrete without worrying about what is inside?
> 
> Hm, the [doc](https://docs.google.com/document/d/1eWW8MTko3PIqxZ32-GhsdaRSYqoDicxMB1VeessMTOg/edit?pli=1&tab=t.0) says of `Action`s: "Whether a constant or type is template-dependent is tracked on its constant value." So I guess TypeOfInst can have an action that is not a template.
> 
> Perhaps I just need to look at its value and see if it's template/symbolic or not.

Well, I turned it into some TODOs. A template-dependent type in impl lookup is weird, since we want to wait to use the type until we know it, so that generates a new TODO in the semir. And a concrete TypeOfInst (I think think this would happen in the specific?) needs some work - I left a TODO saying what I think should be happening, which is to run the action and get an inst back and use its value in the type structure, but I am not totally sure if that makes sense/will be possible/would have side effects..

danakj committed:
no-concrete-type-of-inst

danakj commented:
> Well, I turned it into some TODOs. A template-dependent type in impl lookup is weird, since we want to wait to use the type until we know it, so that generates a new TODO in the semir. And a concrete TypeOfInst (I think think this would happen in the specific?) needs some work - I left a TODO saying what I think should be happening, which is to run the action and get an inst back and use its value in the type structure, but I am not totally sure if that makes sense/will be possible/would have side effects..

I tried adding a call to a template function in `unimplemented.carbon`:
```
class C {
  var n: i32;
}
fn G() {
  F({.n = 1} as C);
}
```

Which results in a specific `F` for `C` being created. `%.loc14_11.5` here is the value of the `type_of_inst` instruction in the generic eval block. The `type_of_inst` does not appear in the generic function body, so I think TypeStructure should not need to worry about concrete `type_of_inst`. I have adjusted it accordingly.
```
// CHECK:STDOUT: specific @F(constants.%C) {
// CHECK:STDOUT:   %T.loc5_15.2 => constants.%C
// CHECK:STDOUT:   %T.patt.loc5_15.2 => constants.%C
// CHECK:STDOUT:
// CHECK:STDOUT: !definition:
// CHECK:STDOUT:   %require_complete => constants.%complete_type.54b
// CHECK:STDOUT:   %.loc14_11.3 => constants.%inst.as_compatible
// CHECK:STDOUT:   %.loc14_11.4 => constants.%inst.splice_block
// CHECK:STDOUT:   %.loc14_11.5 => constants.%i32
// CHECK:STDOUT: }
```

danakj commented:
Josh pointed out to me that we could have a `template` value with a concrete type, like a `template` value of type `i32` or something. At which point the actions on the value would be template-dependent, but `type_of_inst` would not need to be. So question for you @zygoloid is:
- In this case would `type_of_inst` be template dependent anyway because the action instruction it's pointing to is template-dependent, even though its output instruction's type will not be?
- Or in this case would we generate `type_of_inst` at all, or would the type of the `splice_inst` just be the concrete type of which we know the action instruction will generate?

danakj commented:
> Josh pointed out to me that we could have a `template` value with a concrete type, like a `template` value of type `i32` or something. At which point the actions on the value would be template-dependent, but `type_of_inst` would not need to be. So question for you @zygoloid is:
> 
> * In this case would `type_of_inst` be template dependent anyway because the action instruction it's pointing to is template-dependent, even though its output instruction's type will not be?
> * Or in this case would we generate `type_of_inst` at all, or would the type of the `splice_inst` just be the concrete type of which we know the action instruction will generate?

It looks like the _current_ answer is `type_of_inst` is both generated and is template dependent even if the type can be determined concretely.

```
class C {
  var n: i32;
}

// Check that we get a reasonable diagnostic for an unimplemented operation on a template dependent value.
fn F(template c:! C) -> i32 {
  // CHECK:STDERR: fail_todo_unimplemented_value.carbon:[[@LINE+4]]:10: error: cannot access member of interface `Core.Mul` in type `<dependent type>` that does not implement that interface [MissingImplInMemberAccess]
  // CHECK:STDERR:   return c.n * 3;
  // CHECK:STDERR:          ^~~~~~~
  // CHECK:STDERR:
  return c.n * 3;
}

// CHECK:STDOUT: generic fn @F(%c.loc9_15.1: %C) {
// CHECK:STDOUT:   %c.loc9_15.2: %C = bind_symbolic_name c, 0, template [template = %c.loc9_15.2 (constants.%c)]
// CHECK:STDOUT:   %c.patt.loc9_15.2: %C = symbolic_binding_pattern c, 0, template [template = %c.patt.loc9_15.2 (constants.%c.patt)]
// CHECK:STDOUT:
// CHECK:STDOUT: !definition:
// CHECK:STDOUT:   %.loc18_11.2: <instruction> = access_member_action %c.ref, n [template]
// CHECK:STDOUT:   %.loc18_11.3: type = type_of_inst %.loc18_11.2 [template]
// CHECK:STDOUT:
// CHECK:STDOUT:   fn(%c.patt.loc9_15.1: %C) -> %i32 {
// CHECK:STDOUT:   !entry:
// CHECK:STDOUT:     %c.ref: %C = name_ref c, %c.loc9_15.1 [template = %c.loc9_15.2 (constants.%c)]
// CHECK:STDOUT:     %.loc18_11.1: @F.%.loc18_11.3 (@F.%.loc18_11.3) = splice_inst %.loc18_11.2
// CHECK:STDOUT:     %int_3: Core.IntLiteral = int_value 3 [concrete = constants.%int_3]
// CHECK:STDOUT:     return <error>
// CHECK:STDOUT:   }
// CHECK:STDOUT: }
```

danakj committed:
add template value test and more todo text

zygoloid commented on toolchain/check/type_structure.cpp:
```diff
@@ -281,6 +287,36 @@ class TypeStructureBuilder {
+          // CARBON_FATAL code path. For now it is possible, though, to reach
+          // here and it results in a diagnostic.
+          //
+          // However, TypeOfInst can be determined to be a concrete value for a
+          // template value with a non-template dependent type. This currently
+          // does not happen though, the TypeOfInst is always template
+          // dependent. If it does in the future then we will want to use the
+          // concrete constant value instruction of `inst_id` in the type
+          // structure:
+          //   if (const_id.is_concrete()) {
+          //     PushInstId(context_.constant_values().GetInstId(const_id));
+          //   }
```
I think we only see canonical constant type values here, and a `TypeOfInst` instruction will only be a canonical constant value if it's a template constant. (If its argument is an `InstValue`, `TypeOfInst` evaluates to that instruction's value; otherwise it itself defines a template constant.)

I think we probably want to treat a `TypeOfInst` as a `?` in the type structure, because it represents an unknown generic type, and should be a worse match than a known concrete constant type.

danakj replied:
Ok. I have made it so and removed the od TODO. I do think there's some room here for `template n:! i32` where I think you're saying the `TypeOfInst` would be `template` still because its argument is, but the *type* is not template-dependent here so we could look for that and avoid using `?` in that case?

I added a more narrow TODO pointing back to here.

danakj committed:
TypeOfInst is a ? in the type structure

danakj committed:
Merge remote-tracking branch 'origin/trunk' into specialize

danakj committed:
autoupdate
